### PR TITLE
adds waiting time for UI server startup

### PR DESCRIPTION
**Pull request by kotharironak**
This PR will help as an immediate fix for checking the bootstrapping status process. 
- Provides an immediate fix for the issue -  https://github.com/hypertrace/hypertrace/issues/36
- will also allow for configuring wait time for bootstrapping process by start_period, interval, timeout, and retries
- the configuration is also exposed as ENVs

As mentioned in the issue,
- will first start grpc server
- wait for bootstrapping to finish for a configurable time
- will start UI server after configured timeout

### CODE REVIEW

File: hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/HypertraceUIServer.java [hunk 1]
```diff
@@ -90,10 +93,20 @@ private Resource getBaseResource() {
     }
   }
 
+  public void startWithTimerTask() {
+    String defaultTenant = graphQlServiceAppConfig.getString(DEFAULT_TENANT_ID_CONFIG);
+    HypertraceUIServerTimerTask timerTask = new HypertraceUIServerTimerTask(appConfig,
+            this, defaultTenant);
+    LOGGER.info(String.format("Staring a timer task for checking health for bootstrapping process, " +
```
Comment (jcchavezs) [inline hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/HypertraceUIServer.java:100]:
```suggestion
    LOGGER.info(String.format("Starting a timer task for checking health for bootstrapping process, " +
```

Comment (jcchavezs) [reply hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/HypertraceUIServer.java:100]: Log level is controversial but I see this more a debug log.

File: hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/HypertraceUIServerTimerTask.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  public long getInterval() {
+    return interval;
+  }
+
+  @Override
+  public void run() {
+    try {
+      if (numTries >= maxTries) {
```
Comment (jcchavezs) [inline hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/HypertraceUIServerTimerTask.java:71]: I think `numRetries` and `maxRetries` is the common wording?

File: hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/FederatedService.java [hunk 3]
```diff
@@ -133,7 +133,7 @@ protected void doStart() {
     });
 
     Thread uiThread = new Thread(() -> {
-      hypertraceUIServer.start();
+      hypertraceUIServer.startWithTimerTask();
```
Comment (jcchavezs) [inline hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/FederatedService.java:136]: I am curious about this, I don't see `hypertraceUIServer.start` being invoked inside `hypertraceUIServer.startWithTimerTask`.

**Review by adriancole — APPROVED** [review]
This is a good workaround. Later today I will write up what seems to be the bigger issue which makes you have to do this.

For example, the UI health depends on a ready Query layer. A composite service should return health when all of its components are healthy.

While it is possible to use internal loops, also /health invocations can trigger things instead. This is how we do it in zipkin for example. When the state is not healthy, it can re-evaluate that on demand. In this case it boils down to whether or not the UI is handling lack of health well or not. For example, if the UI code re-presented an error initializing, the loop could move there and just hit /health until it appears ready. In other words the spin-loop only relevant to the UI could be better placed in the UI if we allow the UI to load before the service is ready.

Another thing bigger which I'll doc today, is that once health-checks trigger things like schema verification (similar to how zipkin does it), it may well be better to pull UI back out if the UI is unable to spin. That way it won't load until the query layer is in a ready state (storage is up and minimal schema is in). That's an ever better first time experience than a spin loop especially considering when we have only 3 services [hypertrace-ingester, hypertrace-query, hypertrace-ui] it will still be better than now, and also that a 2 process arch [hypertrace-ingester, hypertrace] where the main process is hacky is worse than a 3 process arch. Rationale is that it is still not a single process arch anyway and reliability given by separating the UI part is better than a kindof busted query+ui process.


Anyway, this part is fine to merge for now! Just wanted to give a sneak preview

**Review by JBAhire — APPROVED** [review]

File: hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/HypertraceUIServerTimerTask.java (hunk 2 above)
Comment (kotharironak) [reply hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/HypertraceUIServerTimerTask.java]: Addressed.

File: hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/FederatedService.java (hunk 3 above)
Comment (kotharironak) [reply hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/FederatedService.java:136]: `uiServer.start` will be called from `HypertraceUIServerTimerTask`

**Review by JBAhire — APPROVED** [review]

**Comment by kotharironak** [comment]
@adriancole Let's discuss on offline for 3 service proposal.

File: hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/HypertraceUIServerTimerTask.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  private final GatewayServiceBlockingStub client;
+  private int numRetries;
+  private int maxRetries;
+  private int timeout;
+  private long startTime;
+  private long interval;
+  private long startPeriod;
+  private String defaultTenant;
+
+  public HypertraceUIServerTimerTask(Config appConfig, HypertraceUIServer uiServer, String defaultTenant) {
```
Comment (adriancole) [inline hypertrace-federated-service/src/main/java/org/hypertrace/federated/service/HypertraceUIServerTimerTask.java:43]: fine to merge this, but using this on-demand on health is easier and less code to test as the "scheduling" is external. less timeout config to revert later. in any case we can revert this later.

**Pull request merged**